src/Forms/Form.php

```diff
@@ -317,6 +322,58 @@ public function __construct(
         $this->setupDefaultClasses();
     }
 
+    /**
+     * Make the form require sudo mode, which will make the form readonly and add a sudo mode password field
+     * unless the current user previously activated sudo mode.
+     *
+     * Note that if the parent request handler for this form isn't LeftAndMain or GridFieldDetailForm_ItemRequest,
+     * sudo mode will not be required by this form.
+     */
+    public function requireSudoMode(): void
+    {
+        // Check that the current request handler for the form is one that's used
+        // in an admin context where sudo mode makes sense
+        $classes = [
+            LeftAndMain::class,
+            GridFieldDetailForm_ItemRequest::class,
+        ];
+        $enableSudoMode = false;
+        foreach ($classes as $class) {
+            if (is_a($this->getController(), $class)) {
+                $enableSudoMode = true;
+                break;
+            }
+        }
+        if (!$enableSudoMode) {
+            return;
+        }
+        // Check if sudo mode is currently enabled
+        $service = Injector::inst()->get(SudoModeServiceInterface::class);
+        $session = $this->getRequest()->getSession();
+        if ($service->check($session)) {
+            return;
+        }
+        // If sudo mode is not active, make the form readonly and add a sudo mode password field
+        // Modify the GridFieldViewButton on any GridFields so that it doesn't suffix the view URL with 'view'
+        // This allows us to gracefully reload a form in readonly mode when sudo mode is activated
+        foreach ($this->Fields()->flattenFields() as $field) {
```

Comment:
Is there a way we can do this without calling `flattenFields()`?
That method results in creating a new field list and adding the fields to it, which results in calling `setContainerFieldList()` on the fields, setting the wrong fieldlist for them.

E.g. could use `recursiveWalk()` probably.

Reply:
Moving code to `GridField::Fieldholder()`

Reply:
This is still here?